### PR TITLE
Android - Warn and Patch

### PR DESCRIPTION
**Pull request by cnorthrop**
Okay, I think this is ready for initial review.  +@michaelrgb for Android coverage.

This series does two things:
1. Add a warning when the selected Android application does not contain the RenderDoc layer
2. When clicked, attempt to pull, patch, and reinstall the target application

The warning is modeled after the desktop version.  Both are displayed on my systems (I haven't registered the desktop layer).

warning

For now, the patch offer has a giant disclaimer and suggestion to use Android Studio to get required components.  We can change this as we figure out the right way to do it.  I need to list dependencies (and how to get them) on the wiki.  The patch sequence checks for dependencies first.  I may have missed some, needs to be tested on clean systems.

patch

I'm not worried about the components that are available in AOSP (we can ship those with RenderDoc).  But apksigner uses java.  Need to figure out if shipping a JRE is the answer, or direct people to install it via Android Studio.

One big change is renaming `librenderdoc.so `to `libVkLayer_GLES_RenderDoc.so.`  This allows it to be used with Vulkan and GLES, without renaming (by hand) for local builds.

I haven't tested the series with a GLES app, but with the layer rename, the warning should fire.

I'm sure there is some polishing to be done, and I'm certainly flexible on the text of the popups, but things are functioning well on Windows (with Qt UI), Linux and macOS.

Let me know what you think!

### CODE REVIEW

**Comment by michaelrgb** [comment]
I havent tested this, but:

I do believe that auto-patching APKs just isn't reliable enough to add as a feature inside the GUI. When this breaks on certain games, the user will blame RenderDoc instead of figuring out what they did wrong with their own setup. This will **reflect very badly** on the GUI, and breaks the RenderDoc philosophy of the GUI being "seamless".

Also, this auto-patching will only support Vulkan. For GLES there are no layers so you still need to make the app link to librenderdoc.so (by patching the app's binaries or recompiling). The user will auto-patch his GLES APK and then wonder why he cant capture. Again he will blame the GUI.

Also why change the name of librenderdoc.so inside the server APK? It is not functioning as a layer or hook for the server, it is just a supporting library.

"attempt to pull, patch, and reinstall"
You will have to uninstall the existing app because of different key signatures. A lot of popular games have large data files which they download on first use, which will then need to be re-downloaded.

Unless youre certain this will always work, youre going to get lots of support queries. I really think this is a bad idea to have in the GUI, and we should be trying to figure out a proper supported way of adding debug layers into Android.

**Comment by cnorthrop** [comment]
I can certainly split this up into different pull requests: one to warn, one to patch.  I think the former should be valuable for either API.  Would you be willing to try that out for GLES?

The rename of the *.so was an attempt to unify what to search for when detecting if the layer is present.  Also trying to fix the local build problem.  I might have screwed it up, but the changes I applied were necessary for proper functionality.  I'll look closer at that.

Definitely this won't always work, and it only fixes one potential problem with the APK.  It can't fix the required permissions, for instance.

It would be nice to know if the patch effort helps people or just confuses them further.  Maybe we need to fire up some usability studies. :)  I put it together as a way to help people along, providing clear progress and log entries to show what's wrong with the APK.  I could go farther and print those messages in the progress bar.  Or perhaps we shift gears and just make this an analysis routine?  i.e. Click here to learn what else is wrong with your APK, but don't try to fix it?

FWIW, there will always be scenarios where RenderDoc can't debug an application on production devices, even if the layers are present.  That's how the operating system is designed, and I think developers prefer it that way.  If we shouldn't attempt to fix things unless we can fix everything, then we shouldn't do the patching.

**Comment by michaelrgb** [comment]
How about instead of patching APKs, we push the layer into the data/data/package/lib folder? As you've said, there will always be scenarios where we cant capture on production devices, so lets just require/recommend this feature to be used on engineering devices, so that we are always allowed to push to the app's libraries. Does that sound Ok?

These days I never repackage anyone's APK, I just adb push the library on an engineering device. Its the most reliable to do it.

This way the GUI can rename librenderdoc.so to libVkLayer_GLES_RenderDoc.so when it does adb push, and you dont need to change the java/makefiles.

Yes I like the idea of warning if the layer is not present - that also becomes easier if you check the data/data/.../lib folder.

To try this out for GLES, wouldn't it also need to patch the app's binaries to link to the hook .so? I.e. the same thing that my libpatch.py script does.

**Comment by cnorthrop** [comment]
To clarify, this series is only valuable to GLES in the warning phase.  GLES support can't be patched in since it must be linked against, not simply present in the APK.  I can make this clearer in the disclaimer (that it will only help Vulkan apps).  There is no happy quick fix for that path.

Yes, when running as root things are much easier, but I don't think very many developers have access to engineering builds.  This patch series does check the application's lib directory before warning (`/data/data/<package>/lib` is a symlink to where this checks: `/data/app/<package>/lib/<abi>`, which is readable on production devices).  It also checks `/data/local/debug/vulkan`, which is the [official place](https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/blob/master/loader/LoaderAndLayerInterface.md) (see Android Layer Discovery) to put vulkan layers you want to load on engineering devices.  So if you've already placed layers on the device, the warning doesn't even show up and you're good to go.  Although as I type this, I realize if you are loading a GLES app and you've placed vulkan layers on the system, the warning won't show up.  That's not going to be the common case though.

Part of the rename is to allow RenderDoc to inspect the APK before it knows which graphics API is going to be debugged.  If the file name and location are the same, easy win.

**Comment by cnorthrop** [comment]
Btw, I like the idea of taking the simple route on rooted devices if that's what the user is on, worth some research and support.

Also, I'll spend some time trying to return the error messages when attempting to patch, so it is very clear what's going wrong, rather than just pointing to the log.  Maybe also make the buttons themselves say "Vulkan" and "OpenGL ES" so no one ever tries to patch their GLES app.

**Comment by baldurk** [comment]
I think the auto-patching is definitely a good idea and should be merged. There can certainly be improvements but the concept itself shouldn't be rejected. Android is painful to work with so anything we can do to try and reduce that pain is a good thing.

From my understanding there are two places it falls down:

1. On GLES you have to include renderdoc at link time, so it is impossible to patch in after the fact. We can still detect and warn, but cannot fix. There's a chance slight false-negative here, if the library is present in the apk but not linked properly, but that's probably impossible to detect.
2. Changing the key the APK is signed with and reinstalling can have a variety of impacts - some will be entirely unaffected, others will redownload data, others will completely refuse to run or will break in unexpected ways when authentication/integrity checks fail.

Are there any other places where it can fail? Other than those inherent flaws it seems like basically just manipulating a zip.

> Unless youre certain this will always work, youre going to get lots of support queries. I really think this is a bad idea to have in the GUI, and we should be trying to figure out a proper supported way of adding debug layers into Android.

This solution won't always work but it will save some tedious and error-prone manual intervention in some cases. When it fails it's no worse than doing nothing since the same steps still have to be taken. The process is entirely up-front that the patching is not perfect and that it will re-sign and reinstall the APK. Those developers that might be affected by it should understand the impact.

Also since this is Android, pragmatically we already have our baseline support that we can rely on. Even if a better solution came out tomorrow, we would still have to handle the current level of support for a long time. Maybe in 5 years time we can revisit this once the better solution becomes widespread, but we have to deal with the state of affairs today.

Plus if a new better solution did appear, it could be layered on top of these, just as we could in theory with a solution for rooted phones. We'd try each solution in descending order of preference, from most seamless to least, with this auto-patching being a last option before being forced to ask the user to fix it themselves.

> When this breaks on certain games, the user will blame RenderDoc instead of figuring out what they did wrong with their own setup. This will reflect very badly on the GUI, and breaks the RenderDoc philosophy of the GUI being "seamless".

This assumes a willful ignorance on the part of the user, that after being informed of the process and that it might not work, they'd assume the worst. If someone just clicks through all the dialogs without reading them then they would be equally baffled when clicking through any warnings without patching their APK and seeing that the capture doesn't work. It's also entirely backwards - trying to get things working smoothly for some people is _following_ the philosophy of being "seamless" as far as possible under the circumstances.

If it were entirely up to RenderDoc to define this process, then this all would be done entirely reliably and with at most one prompt for authorisation, exactly the same way as it does on desktop. Unfortunately we are building on top of a bad platform and we are hamstrung by the flaws of Android itself. Trying to mitigate those flaws where possible is still an improvement though. When given a choice between a moderately bad user experience and a strictly worse user experience, choosing the latter because the former isn't perfect is not really helping anyone.

> How about instead of patching APKs, we push the layer into the data/data/package/lib folder? As you've said, there will always be scenarios where we cant capture on production devices, so lets just require/recommend this feature to be used on engineering devices, so that we are always allowed to push to the app's libraries. Does that sound Ok?

We cannot optimise for or require rooted phones which are not typically available among developers. It might make our lives easier to only consider this case but it doesn't reflect the common case that people are going to deal with.

The bottom line for me is that for some people this will make android in renderdoc much more closer to a "just work"ing process rather than having to go read documentation and run some brittle commands. For those people who can't use the auto-patching for one reason or another we're not making things worse, they will still get a friendly warning informing them that the capture won't work due to a missing library or permission, and they'll be directed to the documentation.

File: renderdoc/os/posix/posix_process.cpp [hunk 1]
```diff
@@ -41,14 +41,16 @@ int GetIdentPort(pid_t childPid);
...
+
   // If the application name contains a slash character convert it to an absolute path and return it
   if(appName.find("/") != string::npos)
   {
     char realpathBuffer[PATH_MAX];
     string appDir = dirname(appName);
     string appBasename = basename(appName);
     realpath(appDir.c_str(), realpathBuffer);
-    string appPath(realpathBuffer);
+    appPath.assign(realpathBuffer, strlen(realpathBuffer));
```
Comment (cnorthrop) [inline renderdoc/os/posix/posix_process.cpp]: It didn't occur to me that you could assign a char* to a string like that.  Sure enough, it is supported by operator=.  I'll update it, thanks.

File: renderdoc/os/posix/posix_process.cpp [hunk 2]
```diff
@@ -57,19 +59,28 @@ static const string GetAbsoluteAppPathFromName(const string &appName)
...
-  const char *path = strtok(pathEnvVar, pathSeparator);
+  const char *path = strtok(localPath, pathSeparator);
   while(path)
   {
     string testPath(path);
     testPath += "/" + appName;
     if(!access(testPath.c_str(), X_OK))
-      return testPath;
+    {
+      appPath.assign(testPath);
```
Comment (cnorthrop) [inline renderdoc/os/posix/posix_process.cpp]: Will update.

File: scripts/dist.sh [hunk 3]
```diff
@@ -61,7 +61,7 @@ if [ -f bin-android32/RenderDocCmd.apk ]; then
 	mkdir -p dist/Release64/android/apk dist/Release64/android/lib/armeabi-v7a
 
 	cp bin-android32/RenderDocCmd.apk dist/Release64/android/apk
-	cp bin-android32/librenderdoc.so dist/Release64/android/lib/armeabi-v7a/libVkLayer_RenderDoc.so
+	cp bin-android32/librenderdoc.so dist/Release64/android/lib/armeabi-v7a/libVkLayer_GLES_RenderDoc.so
```
Comment (cnorthrop) [inline scripts/dist.sh]: Good catch, I didn't test the dist changes.  Fixed.

File: qrenderdoc/Windows/Dialogs/CaptureDialog.cpp [hunk 4]
```diff
@@ -397,6 +426,74 @@ void CaptureDialog::vulkanLayerWarn_mouseClick()
...
+      tr("In order to debug on Android, the RenderDoc layer must be present in the "
+         "installed application.<br><br>"
+         "To fix this, you should repackage the APK following guidelines on the "
+         "<a href='http://github.com/baldurk/renderdoc/wiki/Android-Support'>"
+         "RenderDoc Wiki</a><br><br>"
+         "If you are only targeting Vulkan, RenderDoc can try to add the layer for you, "
+         "which requires pulling the APK, patching it, and reinstalling with a debug key. "
+         "This works for many debuggable applications, but not all, especially those that "
+         "check their integrity before launching.<br><br>"
+         "Your system will need Android Studio with several subcomponents installed for "
```
Comment (cnorthrop) [inline qrenderdoc/Windows/Dialogs/CaptureDialog.cpp]: No, it isn't required, I'll update the text.  It is just the easiest way to get all the dependencies.

Although the IDE does create the debug keystore I'm checking for in default location.  If it isn't there, we can certainly create it using keytool.  I might do that as a follow up commit instead of adding it to this one.

File: qrenderdoc/Windows/Dialogs/CaptureDialog.cpp [hunk 5]
```diff
@@ -397,6 +426,74 @@ void CaptureDialog::vulkanLayerWarn_mouseClick()
...
+
+  if(prompt == QMessageBox::Yes)
+  {
+    float progress = 0.0f;
+    bool patchSucceeded = false;
+
+    // call into APK pull, patch, install routine, then continue
+    LambdaThread *patch = new LambdaThread([this, exe, &patchSucceeded, &progress]() {
+      QByteArray hostnameBytes = m_Ctx.Replay().CurrentRemote()->Hostname.toUtf8();
+      if(RENDERDOC_AddLayerToAndroidPackage(hostnameBytes.data(), exe.toStdString().c_str(),
```
Comment (cnorthrop) [inline qrenderdoc/Windows/Dialogs/CaptureDialog.cpp]: Yep, I'll clean this up too.

File: qrenderdoc/Windows/Dialogs/CaptureDialog.cpp [hunk 6]
```diff
@@ -434,7 +531,17 @@ void CaptureDialog::on_exePathBrowse_clicked()
   }
 
   if(!filename.isEmpty())
+  {
     SetExecutableFilename(filename);
+
+    if(m_Ctx.Replay().CurrentRemote() &&
+       m_Ctx.Replay().CurrentRemote()->Hostname.startsWith(lit("adb:")))
```
Comment (cnorthrop) [inline qrenderdoc/Windows/Dialogs/CaptureDialog.cpp]: Ah, thanks.

File: qrenderdoc/Windows/Dialogs/CaptureDialog.cpp [hunk 7]
```diff
@@ -434,7 +531,17 @@ void CaptureDialog::on_exePathBrowse_clicked()
...
+    {
+      QByteArray hostnameBytes = m_Ctx.Replay().CurrentRemote()->Hostname.toUtf8();
+      ui->androidLayerWarn->setVisible(
+          !RENDERDOC_CheckAndroidLayer(hostnameBytes.data(), filename.toStdString().c_str()));
```
Comment (cnorthrop) [inline qrenderdoc/Windows/Dialogs/CaptureDialog.cpp]: Yeah, especially as it gets more expensive to do the check (i.e. permissions).  I'll make it asynchronous and clean up the string handling.

File: renderdoc/replay/entry_points.cpp [hunk 8]
```diff
@@ -573,7 +573,23 @@ void extractDeviceIDAndIndex(const string &hostname, int &index, string &deviceI
...
+  size_t firstSpace = cmd.find(" ");
+  string exe = cmd.substr(0, firstSpace);
+  string args = cmd.substr(firstSpace + 1, cmd.length());
+
+  Process::ProcessResult result;
+  Process::LaunchProcess(exe.c_str(), workDir.c_str(), args.c_str(), &result);
+  if(result.strStdout.length())
+    RDCLOG("STDOUT:\n%s", result.strStdout.c_str());
+  if(result.strStderror.length())
+    RDCLOG("STDERR:\n%s", result.strStderror.c_str());
```
Comment (cnorthrop) [inline renderdoc/replay/entry_points.cpp]: Okay.  This was useful during development, but probably doesn't need to be so verbose.  In the future I think it would be good to return the exit code of commands, but I didn't see a good way to do that right now.

File: renderdoc/replay/entry_points.cpp [hunk 9]
```diff
@@ -678,8 +679,318 @@ uint32_t StartAndroidPackageForCapture(const char *host, const char *package)
...
+  // Walk through the output.  If it starts with META-INF, remove it.
+  std::istringstream contents(fileList);
+  string line;
+  string prefix("META-INF");
+  while(std::getline(contents, line))
+  {
+    line = trim(line);
+    if(line.compare(0, prefix.size(), prefix) == 0)
+    {
+      RDCLOG("Match found, removing  %s", line.c_str());
```
Comment (cnorthrop) [inline renderdoc/replay/entry_points.cpp]: Okay, can do.

File: renderdoc/replay/entry_points.cpp [hunk 10]
```diff
@@ -678,8 +679,318 @@ uint32_t StartAndroidPackageForCapture(const char *host, const char *package)
...
+      execCommand("aapt remove " + apk + " " + line);
+    }
+  }
+
+  // Ensure no hits on second pass through
+  RDCLOG("Walk through file list again, ensure signature removed");
```
Comment (cnorthrop) [inline renderdoc/replay/entry_points.cpp]: Okay.

File: renderdoc/replay/entry_points.cpp [hunk 11]
```diff
@@ -678,8 +679,318 @@ uint32_t StartAndroidPackageForCapture(const char *host, const char *package)
...
+  FileIO::CreateParentDirectory(global);
+  FileIO::Copy(layerPath.c_str(), global.c_str(), true);
+
+  Process::ProcessResult result = execCommand("aapt add " + apk + " " + local, tmpDir);
+
+  if(result.strStdout.empty())
+    return false;
+
+  if(!result.strStderror.empty())
+    return false;
```
Comment (cnorthrop) [inline renderdoc/replay/entry_points.cpp]: Yep, I'll add something there.

File: renderdoc/replay/entry_points.cpp [hunk 12]
```diff
@@ -678,8 +679,318 @@ uint32_t StartAndroidPackageForCapture(const char *host, const char *package)
...
+  if(!errOut.empty())
+    return false;
+
+  // Wait until the aligned version exists to proceed
+  uint32_t elapsed = 0;
+  uint32_t timeout = 10000;    // 10 seconds
+  while(elapsed < timeout)
+  {
+    std::ifstream infile(alignedAPK.c_str());
+    if(infile.good())
```
Comment (cnorthrop) [inline renderdoc/replay/entry_points.cpp]: Ah, thanks.

File: renderdoc/replay/entry_points.cpp [hunk 13]
```diff
@@ -678,8 +679,318 @@ uint32_t StartAndroidPackageForCapture(const char *host, const char *package)
...
+// clang-format off
+#if ENABLED(RDOC_WIN32)
+  execCommand(
+      "cmd /C apksigner.bat sign "
+      "--ks " + debugKey + " "
+      "--ks-pass pass:android "
+      "--key-pass pass:android "
+      "--ks-key-alias androiddebugkey " +
+      apk + "\\",
+      workDir);
```
Comment (cnorthrop) [inline renderdoc/replay/entry_points.cpp]: Yes, apksigner is actually a script, so it needed an interpreter on each operating system.  I can look into hiding the implementation details.

I'll rework the arguments and re-enable clang format.

File: renderdoc/replay/entry_points.cpp [hunk 14]
```diff
@@ -678,8 +679,318 @@ uint32_t StartAndroidPackageForCapture(const char *host, const char *package)
...
+bool CheckPatchingRequirements()
+{
+  // check for aapt, zipalign, apksigner, debug key
+  vector<string> requirements;
+  vector<string> missingTools;
+  requirements.push_back("aapt");
+  requirements.push_back("zipalign");
+  requirements.push_back("apksigner");
+  requirements.push_back("java");
+#if ENABLED(RDOC_WIN32)
```
Comment (cnorthrop) [inline renderdoc/replay/entry_points.cpp]: Will do.

File: renderdoc/replay/entry_points.cpp [hunk 15]
```diff
@@ -678,8 +679,318 @@ uint32_t StartAndroidPackageForCapture(const char *host, const char *package)
...
+  for(uint32_t i = 0; i < requirements.size(); i++)
+  {
+    if(execCommand("cmd /C where " + requirements[i] + " 2>nul").strStdout.empty())
+      missingTools.push_back(requirements[i]);
+  }
+
+  if(execCommand("cmd /C IF EXIST " + debugKey + " ECHO debug key foud").strStdout.empty())
+    missingTools.push_back(debugKey);
```
Comment (cnorthrop) [inline renderdoc/replay/entry_points.cpp]: Yep, I'll move to creating the key in temp, removing the need to rely on user's home directory.

File: renderdoc/replay/entry_points.cpp [hunk 16]
```diff
@@ -890,6 +1203,208 @@ extern "C" RENDERDOC_API void RENDERDOC_CC RENDERDOC_StartAndroidRemoteServer(co
...
+  RDCWARN("No RenderDoc layer for Vulkan was found");
+  return false;
+}
+
+string DetermineInstalledABI(const string &deviceID, const string &packageName)
+{
+  RDCLOG("Checking installed ABI for %s", packageName.c_str());
+  string abi;
+
+  string dump = adbExecCommand(deviceID, "shell pm dump " + packageName).strStdout;
```
Comment (cnorthrop) [inline renderdoc/replay/entry_points.cpp]: Yes, I was thinking of adding an option that disables logging for large commands like this.

I should be able to move to execCommand and not print the output to log, a quick workaround.

File: renderdoc/replay/entry_points.cpp [hunk 17]
```diff
@@ -890,6 +1203,208 @@ extern "C" RENDERDOC_API void RENDERDOC_CC RENDERDOC_StartAndroidRemoteServer(co
...
+  // Check known paths for RenderDoc layer
+  string exePath;
+  FileIO::GetExecutableFilename(exePath);
+  string exeDir = dirname(FileIO::GetFullPathname(exePath));
+
+  std::vector<std::string> paths;
+
+#if defined(RENDERDOC_LAYER_PATH)
+  string customPath(RENDERDOC_LAYER_PATH);
+  RDCLOG("Custom layer path: %s", customPath.c_str());
```
Comment (cnorthrop) [inline renderdoc/replay/entry_points.cpp]: Fair enough, I was just trying to mirror the APK location flexibility.  I'm fine with removing it.

File: renderdoc/replay/entry_points.cpp [hunk 18]
```diff
@@ -890,6 +1203,208 @@ extern "C" RENDERDOC_API void RENDERDOC_CC RENDERDOC_StartAndroidRemoteServer(co
...
+  if(!endswith(customPath, layerName))
+  {
+    if(customPath.back() != '/')
+      customPath += "/";
+    customPath += layerName;
+  }
+  paths.push_back(customPath);
+#endif
+
+  // clang-format off
```
Comment (cnorthrop) [inline renderdoc/replay/entry_points.cpp]: Okay, I'll re-arrange.  Sorry, I'm not a huge fan of clang-format. :)  I have an especially hard time with ampersands on the variable name instead of the type.

File: renderdoc/replay/entry_points.cpp [hunk 19]
```diff
@@ -890,6 +1203,208 @@ extern "C" RENDERDOC_API void RENDERDOC_CC RENDERDOC_StartAndroidRemoteServer(co
...
+  paths.push_back(exeDir + "/android/libs/" + abi + "/" + layerName);                                   // Windows install
+  paths.push_back(exeDir + "/../share/renderdoc/android/libs/" + abi + "/" + layerName);                // Linux install
+  paths.push_back(exeDir + "/../../build-android/renderdoccmd/libs/" + abi + "/" + layerName);          // Local build
+  paths.push_back(exeDir + "/../../../../../build-android/renderdoccmd/libs/" + abi + "/" + layerName); // macOS build
```
Comment (cnorthrop) [inline renderdoc/replay/entry_points.cpp]: I believe we'll still need to do the copy, I don't think aapt is smart enough to take in a relative path.  It will mirror the directory structure given as an argument.  I'll experiment with it though.

File: renderdoc/api/replay/renderdoc_replay.h [hunk 20]
```diff
@@ -1435,3 +1435,12 @@ extern "C" RENDERDOC_API void RENDERDOC_CC RENDERDOC_EnumerateAndroidDevices(rdc
 
 DOCUMENT("Internal function for starting an android remote server.");
 extern "C" RENDERDOC_API void RENDERDOC_CC RENDERDOC_StartAndroidRemoteServer(const char *device);
+
+DOCUMENT("Internal function for checking for presence of RenderDoc Vulkan layer on remote device.");
+extern "C" RENDERDOC_API bool RENDERDOC_CC RENDERDOC_CheckAndroidLayer(const char *host,
+                                                                       const char *exe);
```
Comment (cnorthrop) [inline renderdoc/api/replay/renderdoc_replay.h]: Yeah, I'll rename and incorporate the desktop flag flow.  Thanks for pointing that out.

**Comment by cnorthrop** [comment]
Alright, getting this back in my head.  Sorry for the long delay, and thanks for the thorough review.

I'll push an updated branch once I've incorporated all the changes.

File: renderdoc/replay/entry_points.cpp (hunk 19 above)
Comment (baldurk) [inline renderdoc/replay/entry_points.cpp]: Hmm, but if the folder structure is correct in the builds with `lib/` then won't the copy in the temp folder be in an identical structure? I thought you'd just need to run aapt from the appropriate working directory and it would replicate the correct structure. You'd only need the copy if you're changing the directory structure somehow.

File: renderdoc/replay/entry_points.cpp (hunk 17 above)
Comment (baldurk) [inline renderdoc/replay/entry_points.cpp]: It's not a bad idea, because I'm sure there will be distros that want to put the files somewhere else, I just meant it would need to somehow point to a root folder which we expect to contain `lib/<abi>/libVkLayer_GLES_RenderDoc.so` not directly to an ABI library. Since otherwise I'm not sure how you can support multiple ABIs in a custom location.

File: renderdoc/replay/entry_points.cpp (hunk 18 above)
Comment (baldurk) [inline renderdoc/replay/entry_points.cpp]: Yeh clang-format isn't perfect but the alternatives are me nit-picking all the little formatting things manually, or letting it slide completely and ending up with a codebase with mixed space and tabs, inconsistent formatting, etc.

No-one can ever agree on a coding style, but at least as much as possible clang-format takes any arguments away by automating the process :smile:. Plus since it can be so automated it means you can write the code however you feel comfortable with and just apply the format before committing. There's some stuff that I would prefer to do differently but I couldn't represent it in the clang-format file, so I just write my personal style and it gets reformatted when I save.

File: renderdoc/replay/entry_points.cpp (hunk 8 above)
Comment (baldurk) [inline renderdoc/replay/entry_points.cpp]: Yeh it's absolutely a useful debug feature, but I just think the value proposition is starting to swing around. Especially since hopefully this code should all be stable soon and there shouldn't be a need for people to check for the command output. There will be a happy balance somewhere I just think we'll have to change this default to be more judicious.

**Comment by cnorthrop** [comment]
Okay, new version uploaded.  I was able to address all the feedback.  Beyond the smaller changes, here are the bigger changes:

- Move checking dialogue to LambdaThread
- Draw a "Scanning..." status bar during initial inspection
- Incorporate status flags into Android fixup flow, similar to desktop
- Move script execution into OS specific code, removing ifdefs
- Move dependency searching into OS specific code, removing ifdefs
- Generating debug key in temp, stop using Android Studio's
- Make custom layer path point to lib, and expect the ABI to be present.
- Use relative path to working directory when calling "aapt add" to prevent runtime copy (this required moving build output to libs/lib after packing the APK - you'll probably need to update your build scripts)

After selecting an Android app, you'll see a scanning label like so (this scan will also check locations available to rooted devices):
scanning

Once complete, if the app is missing the RenderDoc library or required permissions, you'll see:
warning

If the app is missing both library and permissions, you'll get the following dialogue, which doesn't progress:
missing_lib_and_permissions

If only the lib is missing, the dialogue will be different, offering to patch for Vulkan, with the previous caveats:
missing_lib

If patching succeeds, the warnings will disappear.  If patching fails, the warning will remain.

Thanks, let me know what you think.  Sorry you have to review it all again after a couple weeks.  :\

-C

File: scripts/dist.sh [hunk 21]
```diff
@@ -75,7 +75,7 @@ if [ -f bin-android64/RenderDocCmd.apk ]; then
 	mkdir -p dist/Release64/android/lib/arm64-v8a
 
 	#cp bin-android64/RenderDocCmd.apk dist/Release64/android/apk/64
-	cp bin-android64/librenderdoc.so dist/Release64/android/lib/arm64-v8a/libVkLayer_RenderDoc.so
+	cp bin-android64/librenderdoc.so dist/Release64/android/lib/arm64-v8a/libVkLayer_GLES_RenderDoc.so
```
Comment (cnorthrop) [inline scripts/dist.sh]: Weird, I fixed that at one point.  Must have gotten lost in a rebase.

File: renderdoc/os/win32/win32_stringio.cpp [hunk 22]
```diff
@@ -186,6 +186,48 @@ string GetFullPathname(const string &filename)
...
+  string filePath;
+
+  // Search the PATH directory list for the application (like shell where) to get the absolute path
+  // Return "" if no exectuable found in the PATH list
+  const DWORD size = 65535;
+  char envPath[size];
+  if(GetEnvironmentVariableA("PATH", envPath, size) == 0)
+    return filePath;
+
+  char *next = nullptr;
```
Comment (cnorthrop) [inline renderdoc/os/win32/win32_stringio.cpp]: Okay

File: renderdoc/os/win32/win32_stringio.cpp [hunk 23]
```diff
@@ -186,6 +186,48 @@ string GetFullPathname(const string &filename)
...
+  const char *pathSeparator = ":";
```
Comment (cnorthrop) [inline renderdoc/os/win32/win32_stringio.cpp]: Strange, that was a semicolon at one point.  I think I experimented with putting the string directly in the parameter list, then accidentally switched it when reverting.  Will fix, for sure.

File: qrenderdoc/Windows/Dialogs/CaptureDialog.cpp [hunk 24]
```diff
@@ -60,6 +60,35 @@ static QString GetDescription(const EnvironmentModification &env)
 
 Q_DECLARE_METATYPE(CaptureSettings);
 
+void CaptureDialog::initWarning(RDLabel *warning, bool visible, void (CaptureDialog::*clickFunc)())
+{
+  if(!warning)
+    return;
+
+  if(clickFunc)
+    QObject::connect(warning, &RDLabel::clicked, this, clickFunc);
```
Comment (cnorthrop) [inline qrenderdoc/Windows/Dialogs/CaptureDialog.cpp]: Okay, I'll back off a little bit there and moves some calls out.

**Comment by cnorthrop** [comment]
> One thing I didn't put on a specific line - I don't understand the need to put the libraries under libs/lib/<abi>/libVkLayer_GLES_RenderDoc.so - why do you need the extra libs/ folder? Wouldn't it work just as well to leave the tree structure the same and set the working directory when invoking aapt to the parent build-android/ folder, so that the relative folder structure still contains the lib/<abi>/... that you want?

This has to do with our current build process.  The "android" and "ant" steps expect the libs/<abi> path.  That's why I put the copy after those steps.  Once we move to the latest SDK, we can move them directory to lib/<abi> and skip libs.  I've found it too disruptive to try rolling forward while developing at the same time, so I haven't finished moving to the new build.

> I think we should take a page from Unity's book actually - when building a Unity project it prompts you to browse to the android SDK root folder. We could do the same thing if we detect the tools aren't available, and locate most of them underneath there. We can cache this in the settings the same way as we do for adb's path, for use on future runs.

Ah, okay, that would work, sure.

> The other one which may or may not be valid - if we only need the keytool to generate the debug key, could we just ship a default debug key in case keytool isn't present? If it's only used to just have some key, and it doesn't have to be generated on the same host that you're installing from, this seems like a feasible solution. We can put it next to adb.exe in the distributed build so it's always available if needed.

This is worth some experimentation.  I haven't actually used a debug key on anything other than the system it was created on.  If they can be moved about, then this would work fine.  I can't think of anything wrong with RenderDoc users all signing their APK with the same debug key.

File: scripts/dist.sh (hunk 21 above)
Comment (cnorthrop) [inline scripts/dist.sh]: Actually, no, I just didn't see that there was a separate 64-bit entry.

**Comment by baldurk** [comment]
> This has to do with our current build process. The "android" and "ant" steps expect the libs/ path. That's why I put the copy after those steps. Once we move to the latest SDK, we can move them directory to lib/ and skip libs. I've found it too disruptive to try rolling forward while developing at the same time, so I haven't finished moving to the new build.

Gotcha, that's fine. It sounds like updating the build process will ease a lot of pain :smile:.

**Comment by cnorthrop** [comment]
Okay, new branch pushed incorporating fixes for the silly mistakes and refactored initWarning to simplify it.

**Comment by michaelrgb** [comment]
Hmm Im now getting this when starting qrenderdoc master branch:

```
Error   - Duplicate docstring 'Internal function for checking remote Android package for requirements
' found on method 'renderdoc.str.c_str' - are you missing a DOCUMENT()?
Fatal   - Found errors in python binding docstrings. Please fix!
```

**Comment by baldurk** [comment]
That's my fault, looks like I need to sort out documentation for the rdctype.str class, it's currently inheriting some other docstrings which lead to a duplicate when I added a new member function.
